dataPipelines/gc_neo4j_publisher/neo4j_publisher.py

```diff
@@ -271,6 +277,69 @@ def process_roles(self) -> None:
+        # which can be unraveled by doing a str.split('.') to extract out the chain of hierarchy
+        flat_hierarchy_dict = dict(flatdict.FlatDict(hierarchy_dict, delimiter='.'))
+        all_hierarchy_nodes = []
+        for key in flat_hierarchy_dict.keys():
+            all_hierarchy_nodes.extend(key.split("."))
+        all_hier_nodes = list(set(all_hierarchy_nodes))
+
+        # check to see if the nodes are currently in the graph, for those that aren't (e.g., `United States Constitution`)
+        # create a new node for those
+        hierarchy_nodes_created = 0
+        for hierarchy_node in all_hier_nodes:
+            match_cypher = "OPTIONAL MATCH (n:Entity) " \
+                           "WHERE n.name='" + hierarchy_node + "' OR '" + hierarchy_node + "' in split(n.aliases,';') "\
+                           "RETURN n IS NOT NULL AS Exists"
+            # result = process_query(match_cypher)
+
+            # create the node if the match_cypher above does not match on an existing node
+            if not process_query(match_cypher)[0]['Exists']:
+                create_cypher = "CREATE (n:Entity {name: '" + hierarchy_node + "'})"
+                process_query(create_cypher)
+                hierarchy_nodes_created+=1
+        print(f"Inserted {hierarchy_nodes_created} hierarchy nodes (entity node type)")
+
+        cypher_list = []
+        # flat_hierarchy_dict.keys() looks similar to below:
+        # Constitution.United States Code.President of the United States.Executive Branch.Secretary of Defense...
+        # split('.') on this to create a list and iterate over tuples to create `HAS_AUTHORITY_OVER` relationships
+        for key in flat_hierarchy_dict.keys():
+            hierarchy_nodes = key.split(".")
+            if len(hierarchy_nodes) > 1:
+                hierarchy_relationships = [(hierarchy_nodes[i], hierarchy_nodes[i + 1]) for i in range(len(hierarchy_nodes) - 1)]
+                for hierarchy_relationship in hierarchy_relationships:
+                    # Example of the cypher generated:
+                    # MATCH (a:Entity), (b:Entity)
+                    # WHERE (a.name = 'Constitution' OR 'Constitution' in split(a.aliases, ';'))
+                    # AND (b.name = 'United States Code' OR 'United States Code' in split(b.aliases, ';'))
+                    # MERGE (a)-[r:HAS_AUTHORITY_OVER]->(b)
+                    cypher_list.append("MATCH (a:Entity), (b:Entity) " +
+                                       "WHERE (a.name = '" + hierarchy_relationship[0] + "' OR '" +
+                                       hierarchy_relationship[0] + "' in split(a.aliases, ';')) " +
+                                       "AND (b.name = '" + hierarchy_relationship[1] + "' OR '" +
+                                       hierarchy_relationship[1] + "' in split(b.aliases, ';')) " +
+                                       "MERGE (a)-[r:HAS_AUTHORITY_OVER]->(b)")
+        cypher_list = list(set(cypher_list))
```

Comment:
note:  this deduplicates the cypher... here is an example of the `flat_hierarchy_dict` variable going into line `321`
![Uploading Screen Shot 2022-09-09 at 2.16.36 PM.png…]()